docs/src/developer/design/user_api/find_dependencies.rst

```diff
@@ -19,4 +19,161 @@
+   for the dependency. That name will be tied to that particular invocation of
+   ``find_or_build_dependency`` and will represent that specific package
+   specification. This is explicitly done by creating a traditional CMake target
+   under the provided name.
+
+   - As a corollary, if the user wants to have multiple specifications of a
+     package, then there must be be multiple calls to
+     ``find_or_build_dependency``, each call must provide a unique name.
+   - The user should continue to refer to this package specification via the
+     provided name, e.g., when using it as a dependency for another
+     ``find_or_build_dependency`` call or when specifying the project's
+     assets (e.g., ``cmaize_add_library``)
+
+****************************************
+Find or Build Dependency Function Design
+****************************************
+
+.. _fig_find_or_build_dependency:
+
+.. figure:: assets/find_or_build_dependency.png
+   :align: center
+
+   Overview of the control flow in the ``find_or_build_dependency`` function.
+
+The control flow of ``find_or_build_dependency`` is diagramed in
+:numref:`fig_find_or_build_dependency`. In addressing the :ref:`fbd_cmake_based`
+consideration the inputs to the ``find_or_build_dependency`` function are the
+name of the dependency and any inputs needed to specify the dependency (e.g.,
+version or URL) and any necessary build options. Internally,
+``find_or_build_dependency`` then:
+
+1. Converts the dependency's information into a ``PackageSpecification`` object.
+2. Obtains the active ``CMaizeProject`` object.
+3. Gets the active ``PackageManager`` from the ``CMaizeProject`` object.
+4. Determines if the ``PackageManager`` object has a dependency matching the
+   ``PackageSpecification`` object.
+5. If yes, then the ``PackageManager`` retrieves the already installed
+   dependency which is returned as a ``InstalledTarget`` object.
+6. If no, then the ``PackageManager`` creates a ``BuildTarget`` object which
+   will be built during the build phase.
+7. Finally the target is registered to the active ``CMaizeProject`` object.
+
+In turn, ``find_or_build_dependency`` thus works like a driver managing the
+interactions of the ``CMaizeProject`` object with the ``PackageManager``
+objects. The reliance on a ``PackageManager`` object addresses the
+:ref:`fbd_wrap_package_managers` consideration.
+
+**********
+API Design
+**********
+
+The :term:`API` of the ``find_or_build_dependency`` function must be capable of
+accepting whatever input a ``PackageSpecification`` object may need. This is
+because the ``PackageManager`` will ultimately work with the
+``PackageSpecification``. To that end, the API will accept a number of
+:term:`kwargs` which will be forwarded to the ``PackageSpecification`` and the
+basic API will be:
+
+.. code-block:: CMake
+
+   cmaize_find_or_build_dependency(name_of_dependency <kwargs>)
+
+
+Here ``name_of_dependency`` will be the name assigned to the CMake target, as
```

Comment:
Just to clarify, `name_of_dependency` will only be the actual CMake target name if the dependency target must be built. If it is already installed, then, at least for the CMakePackageManager, the CMake target will probably have a namespace attached as well.

Regardless, users should refer to the target in CMaize as `name_of_dependency`.

Reply:
It's API breaking, but maybe
```.cmake
cmaize_find_or_build_dependency(
    foo
    ...
)
```
should always result in a CMake target `cmaize_foo`? If we build foo, then `cmaize_foo` aliases a target `foo`. If we find an installed foo, then `cmaize_foo` aliases whatever target the installation provides (presumably the namespaced one).

Reply:
The idea is supposed to be that the name of target the user can use is predictable and unlikely to clash with the real target's name.

Reply:
Actually just disregard these comments. We're specifically referring to targets representing dependencies so users shouldn't be tweaking the targets anyways.

Reply:
@zachcran I pushed some verbiage changes see if that addresses this concern.

Reply:
@zachcran Did these changes address your concern?

Reply:
Sorry, I didn't get any notifications about this until you requested my review again just now. I'll take a look!